c++/src/kj/compat/brotli.c++

```diff
@@ -0,0 +1,355 @@
+    if (amount == 0) {
+      KJ_REQUIRE(atValidEndpoint, "brotli compressed stream ended prematurely");
+      return alreadyRead;
+    } else {
+      next_in = buffer;
+      available_in = amount;
+    }
+  }
+
+  byte* next_out = out;
+  size_t available_out = maxBytes;
+  // Check window bits
+  if (firstInput) {
+    firstInput = false;
+    int streamWbits = _::get_brotli_window_bits(next_in[0]);
+    KJ_REQUIRE(streamWbits <= windowBits,
+        "brotli window size too big", (1 << streamWbits));
+  }
+  BrotliDecoderResult result = BrotliDecoderDecompressStream(
+      ctx, &available_in, &next_in, &available_out, &next_out, nullptr);
+  KJ_REQUIRE(result != BROTLI_DECODER_RESULT_ERROR, "brotli decompression failed",
+             BrotliDecoderErrorString(BrotliDecoderGetErrorCode(ctx)));
+
+  atValidEndpoint = result == BROTLI_DECODER_RESULT_SUCCESS;
+  if (atValidEndpoint && available_in > 0) {
+    // There's more data available. Assume start of new content.
+    // Not sure if we actually want this, but there is limited potential for breakage as arbitrary
+    // trailing data should still be rejected. Unfortunately this is kind of clunky as brotli does
+    // not support resetting an instance.
+    BrotliDecoderDestroyInstance(ctx);
+    ctx = BrotliDecoderCreateInstance(nullptr, nullptr, nullptr);
+    KJ_REQUIRE(ctx, "brotli state allocation failed");
+    firstInput = true;
+  }
+
+  size_t n = maxBytes - available_out;
+  if (n >= minBytes) {
+    return n + alreadyRead;
+  } else {
+    return readImpl(out + n, minBytes - n, maxBytes - n, alreadyRead + n);
+  }
+}
+
+BrotliOutputStream::BrotliOutputStream(OutputStream& inner, int compressionLevel, int windowBits)
+    : inner(inner), ctx(compressionLevel, windowBits) {}
+
+BrotliOutputStream::BrotliOutputStream(OutputStream& inner, decltype(DECOMPRESS), int windowBits)
+    : inner(inner), ctx(nullptr, windowBits) {}
+
+BrotliOutputStream::~BrotliOutputStream() noexcept(false) {
+  pump(BROTLI_OPERATION_FINISH);
+}
+
+void BrotliOutputStream::write(const void* in, size_t size) {
+  ctx.setInput(in, size);
+  pump(BROTLI_OPERATION_PROCESS);
+}
+
+void BrotliOutputStream::pump(BrotliEncoderOperation flush) {
+  bool ok;
+  do {
+    auto result = ctx.pumpOnce(flush);
+    ok = get<0>(result);
+    auto chunk = get<1>(result);
+    if (chunk.size() > 0) {
+      inner.write(chunk.begin(), chunk.size());
+    }
+  } while (ok);
+}
+
+// =======================================================================================
+
+BrotliAsyncInputStream::BrotliAsyncInputStream(AsyncInputStream& inner, kj::Maybe<int> _windowBits)
+    : inner(inner), next_in(nullptr), available_in(0) {
+  windowBits = _windowBits.orDefault(_::KJ_BROTLI_MAX_DEC_WBITS);
+  KJ_REQUIRE(windowBits >= BROTLI_MIN_WINDOW_BITS && windowBits <= BROTLI_MAX_WINDOW_BITS,
+      "invalid brotli window size (window bits", windowBits, ")");
+  ctx = BrotliDecoderCreateInstance(nullptr, nullptr, nullptr);
```

Comment:
Nit: `windowBits` and `ctx` could be initialized in the constructor initializer list